### ISSUE TITLE
r10k is profoundly unhelpful by default

### ISSUE THREAD

**Issue by igalic** [puppetlabs/r10k#20]
```
igalic@puppet01 ~ % r10k
/usr/local/bin/r10k:23:in `load': no such file to load -- /var/lib/gems/1.8/gems/r10k-0.0.9/bin/r10k (LoadError)
    from /usr/local/bin/r10k:23
1 igalic@puppet01 ~ % r10k --help
/usr/local/bin/r10k:23:in `load': no such file to load -- /var/lib/gems/1.8/gems/r10k-0.0.9/bin/r10k (LoadError)
    from /usr/local/bin/r10k:23
1 igalic@puppet01 ~ % r10k help
/usr/local/bin/r10k:23:in `load': no such file to load -- /var/lib/gems/1.8/gems/r10k-0.0.9/bin/r10k (LoadError)
    from /usr/local/bin/r10k:23
1 igalic@puppet01 ~ %                                                                      :(
```

**Comment by adrienthebo** [comment]
Well, that's very bad. This is what should normally happen:

<pre>
bin/r10k
NAME
    r10k - Killer robot powered Puppet environment deployment
USAGE
    r10k <subcommand> [options]
DESCRIPTION
    r10k is a suite of commands to help deploy and manage puppet code for complex
    environments.
COMMANDS
    cache           Update cache for all sources
    environment     Operate on a specific environment
    module          Operate on a specific puppet module
    synchronize     Fully synchronize all environments
OPTIONS
    -c --config     Specify a configuration file
    -h --help       show help for this command
    -t --trace      Display stack traces on application crash
    -v --verbose    Set verbosity level
</pre>

I suspect this might be an issue with how `/usr/local/bin/r10k` is loading the ruby gem, or something like that. Could you provide the output of `gem list` and paste the contents of `/usr/local/bin/r10k` ? In addition, what operating system is this?

**Comment by igalic** [comment]
more context: This happens (to me) on two different machines, each with Ubuntu. One has 1.9, the other 1.8 - my shell is zsh - which is the only thing (so far as I can see) that is different from other users' env, where it works out fine.

**Comment by adrienthebo** [comment]
@igalic my understanding is that the failure is happening on the machine using Ruby 1.9. Could you provide the output `gem env` ?

**Comment by igalic** [comment]
```
RubyGems Environment:
  - RUBYGEMS VERSION: 1.8.23
  - RUBY VERSION: 1.9.3 (2012-04-20 patchlevel 194) [x86_64-linux]
  - INSTALLATION DIRECTORY: /var/lib/gems/1.9.1
  - RUBY EXECUTABLE: /usr/bin/ruby1.9.1
  - EXECUTABLE DIRECTORY: /usr/local/bin
  - RUBYGEMS PLATFORMS:
    - ruby
    - x86_64-linux
  - GEM PATHS:
     - /var/lib/gems/1.9.1
     - /home/igalic/.gem/ruby/1.9.1
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :benchmark => false
     - :backtrace => false
     - :bulk_threshold => 1000
  - REMOTE SOURCES:
     - http://rubygems.org/
```

n.b.: ~/.gem/ 

```
igalic@tynix ~ % ls -l .gem
total 4
drwxrwxr-x 3 igalic igalic 4096 Feb  7 17:47 specs/
igalic@tynix ~ %  
```

**Comment by adrienthebo** [comment]
That was unexpected. Could you paste the contents of `/usr/local/bin/r10k` ? The fact that `/var/lib/gems/1.8` is being checked is very odd to me.

**Comment by igalic** [comment]
```
#!/usr/bin/ruby1.9.1
#
# This file was generated by RubyGems.
#
# The application 'r10k' is installed as part of a gem, and
# this file is here to facilitate running it.
#

require 'rubygems'

version = ">= 0"

if ARGV.first
  str = ARGV.first
  str = str.dup.force_encoding("BINARY") if str.respond_to? :force_encoding
  if str =~ /\A_(.*)_\z/
    version = $1
    ARGV.shift
  end
end

gem 'r10k', version
load Gem.bin_path('r10k', 'r10k', version)
```

**Comment by adrienthebo** [comment]
https://github.com/adrienthebo/r10k/issues/20#issuecomment-17035408 indicates that those commands were run on the machine `tynix`, but the original machine was `puppet01`; is all of the above output from the same machine or different machines?

**Comment by igalic** [comment]
Here's the above from `puppet01`:

```
igalic@puppet01 ~ % r10k help
/usr/local/bin/r10k:23:in `load': no such file to load -- /var/lib/gems/1.8/gems/r10k-0.0.9/bin/r10k (LoadError)
        from /usr/local/bin/r10k:23
1 igalic@puppet01 ~ % gem env
RubyGems Environment:
  - RUBYGEMS VERSION: 1.8.24
  - RUBY VERSION: 1.8.7 (2011-06-30 patchlevel 352) [x86_64-linux]
  - INSTALLATION DIRECTORY: /var/lib/gems/1.8
  - RUBY EXECUTABLE: /usr/bin/ruby1.8
  - EXECUTABLE DIRECTORY: /usr/local/bin
  - RUBYGEMS PLATFORMS:
    - ruby
    - x86_64-linux
  - GEM PATHS:
     - /var/lib/gems/1.8
     - /home/rz01/igalic/.gem/ruby/1.8
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :benchmark => false
     - :backtrace => false
     - :bulk_threshold => 1000
  - REMOTE SOURCES:
     - http://rubygems.org/
igalic@puppet01 ~ % ls -l .gem
total 8
drwxr-xr-x 3 igalic rz01 4096 Feb 15 08:39 ruby/
drwxr-xr-x 3 igalic rz01 4096 Feb 15 08:39 specs/
igalic@puppet01 ~ % ls -l .gem/ruby
total 4
drwxr-xr-x 3 igalic rz01 4096 Feb 15 08:39 1.8/
igalic@puppet01 ~ % ls -l .gem/ruby/1.8/cache
total 8
-rw-r--r-- 1 igalic rz01 8192 Feb 15 08:39 ruby-shadow-2.1.4.gem
igalic@puppet01 ~ % cat /usr/local/bin/r10k
#!/usr/bin/ruby1.8
#
# This file was generated by RubyGems.
#
# The application 'r10k' is installed as part of a gem, and
# this file is here to facilitate running it.
#

require 'rubygems'

version = ">= 0"

if ARGV.first
  str = ARGV.first
  str = str.dup.force_encoding("BINARY") if str.respond_to? :force_encoding
  if str =~ /\A_(.*)_\z/
    version = $1
    ARGV.shift
  end
end

gem 'r10k', version
load Gem.bin_path('r10k', 'r10k', version)

igalic@puppet01 ~ % 
```

**Comment by antaflos** [comment]
Seems the problem occurs when running `r10k` as a non-root user:

```
daff@puppet01:~$ whoami 
daff
daff@puppet01:~$ r10k
/usr/local/bin/r10k:23:in `load': no such file to load -- /var/lib/gems/1.8/gems/r10k-0.0.9/bin/r10k (LoadError)
        from /usr/local/bin/r10k:23
daff@puppet01:~$ sudo -s
root@puppet01:~# whoami
root
root@puppet01:~# r10k
NAME
    r10k - Killer robot powered Puppet environment deployment

USAGE
    r10k <subcommand> [options]
...
```

The output of `gem env` is exactly the same when running as root and non-root (daff in this case):

```
RubyGems Environment:
  - RUBYGEMS VERSION: 1.8.24
  - RUBY VERSION: 1.8.7 (2011-06-30 patchlevel 352) [x86_64-linux]
  - INSTALLATION DIRECTORY: /var/lib/gems/1.8
  - RUBY EXECUTABLE: /usr/bin/ruby1.8
  - EXECUTABLE DIRECTORY: /usr/local/bin
  - RUBYGEMS PLATFORMS:
    - ruby
    - x86_64-linux
  - GEM PATHS:
     - /var/lib/gems/1.8
     - /home/rz01/daff/.gem/ruby/1.8
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :benchmark => false
     - :backtrace => false
     - :bulk_threshold => 1000
  - REMOTE SOURCES:
     - http://rubygems.org/
```

**Comment by adrienthebo** [comment]
Well, expletives. Could you do `find /var/lib/gems -name '*r10k*' -ls` and post the permissions here?

**Comment by antaflos** [comment]
Of course:

```
root@puppet01:~# find /var/lib/gems -name "*r10k*" -ls
792027   16 -rw-r--r--   1 root     root        13824 Apr 15 16:45 /var/lib/gems/1.8/cache/r10k-0.0.9.gem
1176193    4 drwxr-xr-x   4 root     root         4096 Apr 17 04:35 /var/lib/gems/1.8/gems/r10k-0.0.9
1176197    4 drwxr-xr-x  10 root     root         4096 Apr 17 04:35 /var/lib/gems/1.8/gems/r10k-0.0.9/lib/r10k
1176238    4 -rw-------   1 root     root          136 Apr 17 04:35 /var/lib/gems/1.8/gems/r10k-0.0.9/lib/r10k.rb
1176195    4 -rwx--x--x   1 root     root          348 Apr 17 04:35 /var/lib/gems/1.8/gems/r10k-0.0.9/bin/r10k
792031    4 -rw-r--r--   1 root     root         1956 Apr 17 04:35 /var/lib/gems/1.8/specifications/r10k-0.0.9.gemspec
1176563    4 drwxr-xr-x   4 root     root         4096 Apr 17 04:35 /var/lib/gems/1.8/doc/r10k-0.0.9
1183198    4 drwxr-xr-x  10 root     root         4096 Apr 17 04:35 /var/lib/gems/1.8/doc/r10k-0.0.9/rdoc/files/lib/r10k
1183268    4 -rw-r--r--   1 root     root         2276 Apr 17 04:35 /var/lib/gems/1.8/doc/r10k-0.0.9/rdoc/files/lib/r10k_rb.html
```

Seems like `/var/lib/gems/1.8/gems/r10k-0.0.9/lib/r10k.rb` is the culprit?

**Comment by igalic** [comment]
After changing my permissions to:

```
igalic@tynix ~ % gem contents  r10k | xargs ls -lahrt
-rw-r--r-- 1 root root  280 May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/cli/ext/logging.rb
-rw-r--r-- 1 root root  886 May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/cli/environment/stale.rb
-rw-r--r-- 1 root root  527 May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/cli/environment.rb
-rw-r--r-- 1 root root  640 May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/cli/environment/list.rb
-rw-r--r-- 1 root root 1.3K May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/cli/environment/deploy.rb
-rw-r--r-- 1 root root  753 May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/cli/cache.rb
-rw-r--r-- 1 root root  104 May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/action.rb
-rw-r--r-- 1 root root  860 May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/action/module.rb
-rw-r--r-- 1 root root 1.9K May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/action/environment.rb
-rwxr-xr-x 1 root root  348 May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/bin/r10k
-rw-r--r-- 1 root root   36 May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/version.rb
-rw-r--r-- 1 root root  460 May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/util/interp.rb
-rw-r--r-- 1 root root 6.0K May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/synchro/git.rb
-rw-r--r-- 1 root root 2.2K May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/root.rb
-rw-r--r-- 1 root root  136 May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k.rb
-rw-r--r-- 1 root root 1.4K May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/module.rb
-rw-r--r-- 1 root root  453 May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/module/git.rb
-rw-r--r-- 1 root root 2.0K May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/module/forge.rb
-rw-r--r-- 1 root root  937 May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/logging.rb
-rw-r--r-- 1 root root  564 May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/librarian.rb
-rw-r--r-- 1 root root  362 May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/librarian/dsl.rb
-rw-r--r-- 1 root root  122 May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/errors.rb
-rw-r--r-- 1 root root  755 May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/deployment.rb
-rw-r--r-- 1 root root 1.9K May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/deployment/environment_collection.rb
-rw-r--r-- 1 root root 1.1K May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/config.rb
-rw-r--r-- 1 root root  747 May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/config/loader.rb
-rw-r--r-- 1 root root 1.1K May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/cli/synchronize.rb
-rw-r--r-- 1 root root 1.1K May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/cli.rb
-rw-r--r-- 1 root root  533 May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/cli/module.rb
-rw-r--r-- 1 root root 1.1K May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/cli/module/list.rb
-rw-r--r-- 1 root root 1.4K May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/r10k/cli/module/deploy.rb
-rw-r--r-- 1 root root 3.9K May 10 13:02 /var/lib/gems/1.9.1/gems/r10k-0.0.9/lib/semver.rb
```

Everything works out fine:

```
igalic@tynix ~ % r10k --help
NAME
    r10k - Killer robot powered Puppet environment deployment

USAGE
    r10k <subcommand> [options]

DESCRIPTION
    r10k is a suite of commands to help deploy and manage puppet code for complex
    environments.

COMMANDS
    cache           Update cache for all sources
    environment     Operate on a specific environment
    module          Operate on a specific puppet module
    synchronize     Fully synchronize all environments

OPTIONS
    -c --config     Specify a configuration file
    -h --help       show help for this command
    -t --trace      Display stack traces on application crash
    -v --verbose    Set verbosity level
igalic@tynix ~ %  
```

**Comment by adrienthebo** [comment]
Expletives. This was a longstanding problem with everything I published because my umask was `0077`. It turns out that gems keep the build time permissions which breaks everything and caused GH-1. What version of r10k was this?

**Comment by igalic** [comment]
0.0.9

**Comment by adrienthebo** [comment]
I'm going to be releasing 1.0.0 Real Soon Now (TM) which should handle the permission bug, so this should be resolved shortly.

**Comment by adrienthebo** [comment]
Well, it meant cutting two RCs, but I finally managed to publish a version of this with sane permissions. This should be fixed in r10k 1.0.0rc2. 

![facepalm](http://cdn.memegenerator.net/instances/400x/37895839.jpg)
